assets/js/atomic/utils/register-block-single-product-template.ts

```diff
@@ -1,46 +1,96 @@
 /**
  * External dependencies
  */
-import { getBlockType } from '@wordpress/blocks';
+import {
+	BlockAttributes,
+	BlockConfiguration,
+	BlockVariation,
+	getBlockType,
+	registerBlockType,
+	registerBlockVariation,
+	unregisterBlockType,
+	unregisterBlockVariation,
+} from '@wordpress/blocks';
 import { subscribe, select } from '@wordpress/data';
 
 export const registerBlockSingleProductTemplate = ( {
-	registerBlockFn,
-	unregisterBlockFn,
 	blockName,
+	blockMetadata,
+	blockSettings,
+	isVariationBlock = false,
+	variationName,
 }: {
-	registerBlockFn: () => void;
-	unregisterBlockFn: () => void;
 	blockName: string;
+	blockMetadata: Partial< BlockConfiguration >;
+	blockSettings: Partial< BlockConfiguration >;
+	isVariationBlock?: boolean;
+	variationName?: string;
 } ) => {
-	let currentTemplateId: string | undefined;
+	let currentTemplateId: string | undefined = '';
 
 	subscribe( () => {
 		const previousTemplateId = currentTemplateId;
 		const store = select( 'core/edit-site' );
-		currentTemplateId = store?.getEditedPostId() as string | undefined;
+		currentTemplateId = store?.getEditedPostContext< {
+			templateSlug?: string;
+		} >()?.templateSlug;
+		const hasChangedTemplate = previousTemplateId !== currentTemplateId;
+		const hasTemplateId = Boolean( currentTemplateId );
 
-		if ( previousTemplateId === currentTemplateId ) {
+		if ( ! hasChangedTemplate || ! hasTemplateId || ! blockName ) {
 			return;
 		}
 
-		const parsedTemplate = currentTemplateId?.split( '//' )[ 1 ];
-
-		if ( parsedTemplate === null || parsedTemplate === undefined ) {
-			return;
-		}
-
-		const block = getBlockType( blockName );
+		let isBlockRegistered = Boolean( getBlockType( blockName ) );
 
 		if (
-			block === undefined &&
-			parsedTemplate.includes( 'single-product' )
+			isBlockRegistered &&
+			( currentTemplateId?.includes( 'single-product' ) ||
+				previousTemplateId?.includes( 'single-product' ) )
 		) {
-			registerBlockFn();
+			if ( isVariationBlock && variationName ) {
+				unregisterBlockVariation( blockName, variationName );
+			} else {
+				unregisterBlockType( blockName );
+			}
+			isBlockRegistered = false;
 		}
```

Comment:
I'm a bit confused by this code. Why is it needed to unregister the blocks here?

Reply:
Makes sense, thanks for explaining, @thealexandrelara. Do you think we could add a comment in that file with part of your explanation so anybody looking at the file knows its purpose?